### PR TITLE
Fix index prefixes to work with span_multi

### LINKED ISSUE #31056: Make index_prefixes work with span_multi
Currently it fails because `span_multi` expects a `MultiTermQuery` while prefix queries might create term queries when `index_prefixes` in enabled. Maybe `span_multi` should also accept `term` queries wrapped in a `constant_score`? There might also be things we need to change regarding how we decide on what index options to use on the prefix field type.

I think it would be a major improvement for users who need to run queries that look like `a NEAR b*` who currently kill their clusters because the `span_multi` expands to too many terms. 

Here is a recreation:

```
PUT index
{
  "mappings": {
    "_doc": {
      "properties": {
        "foo": {
          "type": "text",
          "index_prefixes": {}
        }
      }
    }
  }
}

PUT index/_doc/1
{
  "foo": "the quick fox jumps over the lazy dog"
}

GET index/_search
{
  "query": {
    "span_near": {
      "clauses": [
        {
          "span_multi": {
            "match": {
              "prefix": {
                "foo": {
                  "value": "jumps"
                }
              }
            }
          }
        },
        {
          "span_term": {
            "foo": "fox"
          }
        }
      ],
      "slop": 2
    }
  }
}
```

which triggers:

```
{
  "error": {
    "root_cause": [
      {
        "type": "query_shard_exception",
        "reason": "failed to create query: {\n  \"span_near\" : {\n    \"clauses\" : [\n      {\n        \"span_multi\" : {\n          \"match\" : {\n            \"prefix\" : {\n              \"foo\" : {\n                \"value\" : \"jumps\",\n                \"boost\" : 1.0\n              }\n            }\n          },\n          \"boost\" : 1.0\n        }\n      },\n      {\n        \"span_term\" : {\n          \"foo\" : {\n            \"value\" : \"fox\",\n            \"boost\" : 1.0\n          }\n        }\n      }\n    ],\n    \"slop\" : 2,\n    \"in_order\" : true,\n    \"boost\" : 1.0\n  }\n}",
        "index_uuid": "L67oony8REGtPcAHC0C8WQ",
        "index": "index"
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": "index",
        "node": "MYCVYLGiSiy2tZzASlk7uA",
        "reason": {
          "type": "query_shard_exception",
          "reason": "failed to create query: {\n  \"span_near\" : {\n    \"clauses\" : [\n      {\n        \"span_multi\" : {\n          \"match\" : {\n            \"prefix\" : {\n              \"foo\" : {\n                \"value\" : \"jumps\",\n                \"boost\" : 1.0\n              }\n            }\n          },\n          \"boost\" : 1.0\n        }\n      },\n      {\n        \"span_term\" : {\n          \"foo\" : {\n            \"value\" : \"fox\",\n            \"boost\" : 1.0\n          }\n        }\n      }\n    ],\n    \"slop\" : 2,\n    \"in_order\" : true,\n    \"boost\" : 1.0\n  }\n}",
          "index_uuid": "L67oony8REGtPcAHC0C8WQ",
          "index": "index",
          "caused_by": {
            "type": "unsupported_operation_exception",
            "reason": "unsupported inner query, should be org.apache.lucene.search.MultiTermQuery but was org.apache.lucene.search.ConstantScoreQuery"
          }
        }
      }
    ]
  },
  "status": 400
}
```

cc @jimczi @romseygeek

### PR DESCRIPTION
**Pull request by jimczi**
Text fields that use `index_prefixes` can rewrite `prefix` queries into
`term` queries internally. This commit fix the handling of this rewriting
in the `span_multi` query.
This change also copies the index options of the text field into the
prefix field in order to be able to run positional queries. This is mandatory
for `span_multi` to work but this could also be useful to optimize `match_phrase_prefix`
queries in a follow up. Note that this change can only be done on indices created
after 6.3 since we set the index options to doc only in this version.

Fixes #31056

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java [hunk 1]
```diff
@@ -124,22 +134,61 @@ public static SpanMultiTermQueryBuilder fromXContent(XContentParser parser) thro
     protected Query doToQuery(QueryShardContext context) throws IOException {
         Query subQuery = multiTermQueryBuilder.toQuery(context);
         float boost = AbstractQueryBuilder.DEFAULT_BOOST;
-        if (subQuery instanceof BoostQuery) {
+        if (subQuery instanceof ConstantScoreQuery) {
+            subQuery = ((ConstantScoreQuery) subQuery).getQuery();
+        } else if (subQuery instanceof BoostQuery) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: should we unwrap recursively to be more resilient to changes to the way queries are built?

File: server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java [hunk 2]
```diff
@@ -124,22 +134,61 @@ public static SpanMultiTermQueryBuilder fromXContent(XContentParser parser) thro
...
-            throw new UnsupportedOperationException("unsupported inner query, should be " + MultiTermQuery.class.getName() +" but was "
-                    + subQuery.getClass().getName());
+        if (subQuery instanceof TermQuery) {
+            /**
+             * Text fields that index prefixes can rewrite prefix queries
+             * into term queries. See {@link TextFieldMapper.TextFieldType#prefixQuery}.
+             */
+            if (multiTermQueryBuilder.getClass() != PrefixQueryBuilder.class) {
+                throw new UnsupportedOperationException("unsupported inner query, should be "
+                    + MultiTermQuery.class.getName() + " but was " + subQuery.getClass().getName());
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/SpanMultiTermQueryBuilder.java]: let's include the class of multiTermQueryBuilder to the error message as well?

File: server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java [hunk 3]
```diff
@@ -175,7 +176,11 @@ public TextFieldMapper build(BuilderContext context) {
                 if (fieldType().isSearchable() == false) {
                     throw new IllegalArgumentException("Cannot set index_prefixes on unindexed field [" + name() + "]");
                 }
-                if (fieldType.indexOptions() == IndexOptions.DOCS_AND_FREQS_AND_POSITIONS_AND_OFFSETS) {
+                // Copy the index options of the main field to allow phrase queries on
+                // the prefix field.
+                if (context.indexCreatedVersion().onOrAfter(Version.V_7_0_0_alpha1)) {
+                    prefixFieldType.setIndexOptions(fieldType.indexOptions());
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java]: should we map `DOCS_AND_FREQS` TO `DOCS`?

File: server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java [hunk 4]
```diff
@@ -175,7 +176,11 @@ public TextFieldMapper build(BuilderContext context) {
                 if (fieldType().isSearchable() == false) {
                     throw new IllegalArgumentException("Cannot set index_prefixes on unindexed field [" + name() + "]");
                 }
-                if (fieldType.indexOptions() == IndexOptions.DOCS_AND_FREQS_AND_POSITIONS_AND_OFFSETS) {
+                // Copy the index options of the main field to allow phrase queries on
+                // the prefix field.
+                if (context.indexCreatedVersion().onOrAfter(Version.V_7_0_0_alpha1)) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java]: I *think* this can already be set to 6_4 before the backport, can't it?

**Review by jpountz — APPROVED** [review]
Left some minor comments, looks good in general. Thanks @jimczi.

File: server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java [hunk 5]
```diff
@@ -339,7 +344,6 @@ public Query existsQuery(QueryShardContext context) {
         PrefixFieldType(String name, int minChars, int maxChars) {
             setTokenized(true);
             setOmitNorms(true);
-            setIndexOptions(IndexOptions.DOCS);
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java]: I think we still need this?  What happens with a 6.3 index that doesn't store offsets?

**Review by romseygeek — APPROVED** [review]
I left one comment, but looks good otherwise

**Pull request merged**